### PR TITLE
feat(controller): Delay pod GC deletion by a configurable amount (default 5s)

### PR DESCRIPTION
**Pull request by stefansedich**
The PR https://github.com/argoproj/argo-workflows/pull/5033 was meant to fix the issue of delaying pod deletion for some period so things like logging scrapers can do their thing before the pod goes away.

Based on my comments https://github.com/argoproj/argo-workflows/pull/5033#discussion_r581354568 this won't actually work as intended and another solution is needed.

My fix in this PR aims to delay the actual deletion by some configurable value and the deletion logic will then either delete immediately or queue the deletion for after this period.

@terrytangyuan tagging you as you did the original PR and tagging @alexec as you were the reviewer, as you might have more insight here. I have written an initial test here but am happy to expand coverage if desired.

One area I am not sure about is what other docs need to be updated for an API change and if that is a manual process or if there is some automation around that, it was not clear to me as I could not find any documentation explaining how to make contributions that add to the spec and this is my first time so just let me know what else I need to do.

Checklist:

* [X] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -862,6 +864,11 @@ func (podGC *PodGC) GetLabelSelector() *metav1.LabelSelector {
 	return nil
 }
 
+// GetDelay gets the podGC delay as a duration.
+func (podGC *PodGC) GetDelay() time.Duration {
+	return time.Duration(podGC.DelaySeconds) * time.Second
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:869]: Will it handle the default? what will the delay be if the user didn't specify in Workflow?

Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:869]: `DelaySeconds` should default to `0` and `GetDelay` in this case should return `0 seconds`, meaning the deletion will be queued right away.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -840,6 +840,8 @@ type PodGC struct {
 	Strategy PodGCStrategy `json:"strategy,omitempty" protobuf:"bytes,1,opt,name=strategy,casttype=PodGCStrategy"`
 	// LabelSelector is the label selector to check if the pods match the labels before being added to the pod GC queue.
 	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty" protobuf:"bytes,2,opt,name=labelSelector"`
+	// DelaySeconds is how long to delay pod GC for.
+	DelaySeconds uint32 `json:"delaySeconds,omitempty" protobuf:"bytes,3,opt,name=delaySeconds"`
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:844]: Can you change it to `intstr.IntOrString` some it can be parameterized by parameters?

**Review by sarabala1979 — DISMISSED** [review]

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 2 above)
Comment (stefansedich) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: @sarabala1979 done, hopefully I got that right.

**Comment by alexec** [comment]
I spoke with @sarabala1979 today and we think it is likely that few users will be impacted by a short delay on Pod GC, and many who'll benefit. 

I suggest we have a single global value this suitable default. No per-workflow configuration.

**Comment by stefansedich** [comment]
> I spoke with @sarabala1979 today and we think it is likely that few users will be impacted by a short delay on Pod GC, and many who'll benefit.
> 
> I suggest we have a single global value this suitable default. No per-workflow configuration.

Thanks @alexec this makes sense, will make things simpler and only require some more sleeps in the tests, did you talk about a reasonable default?

And by global default value do you mean configurable with config still?

**Comment by alexec** [comment]
I don't really know what a sensible default would be - maybe 10s? How long were you thinking you'd need? It should be configurable using the `workflow-controller-configmap`.

**Comment by stefansedich** [comment]
> I don't really know what a sensible default would be - maybe 10s? How long were you thinking you'd need? It should be configurable using the `workflow-controller-configmap`.

In our case I planned to configure a default of 1m or so, 10s sounds like a reasonable default however for most and this does simplify things quite a bit, let me get that done.

File: config/config.go [hunk 3]
```diff
@@ -106,12 +107,16 @@ type Config struct {
 	// PodSpecLogStrategy enables the logging of podspec on controller log.
 	PodSpecLogStrategy PodSpecLogStrategy `json:"podSpecLogStrategy,omitempty"`
 
-	// PodGCGracePeriodSeconds specifies the duration in seconds before the pods in the GC queue get deleted.
-	// Value must be non-negative integer. A zero value indicates that the pods will be deleted immediately
-	// as soon as they arrived in the pod GC queue.
-	// Defaults to 30 seconds.
+	// PodGCGracePeriodSeconds specifies the duration in seconds before a terminating pod is forcefully killed.
+	// Value must be non-negative integer. A zero value indicates that the pod will be forcefully terminated immediately.
+	// Defaults to the Kubernetes default of 30 seconds.
```
Comment (stefansedich) [inline config/config.go:112]: I changed this comment to signify what this option does, as the new option below handles the actual deletion delay.

File: config/config.go [hunk 4]
```diff
@@ -144,6 +149,14 @@ func (c Config) GetResourceRateLimit() ResourceRateLimit {
 	}
 }
 
+func (c Config) GetPodGCDeleteDelay() time.Duration {
+	if c.PodGCDeleteDelay == nil {
+		return 30 * time.Second
```
Comment (stefansedich) [inline config/config.go:154]: Defaulting to 30s, happy to lower but this was somewhat in the middle of the suggested 10s and the 1m we would run in our environment.

File: test/e2e/pod_cleanup_test.go [hunk 5]
```diff
@@ -21,7 +21,7 @@ type PodCleanupSuite struct {
 	fixtures.E2ESuite
 }
 
-const enoughTimeForPodCleanup = 3 * time.Second
+const enoughTimeForPodCleanup = 33 * time.Second
```
Comment (stefansedich) [inline test/e2e/pod_cleanup_test.go:24]: Padding out tests as they will need a little more time to "wait" now as there is a default deletion delay.

**Comment by stefansedich** [comment]
@alexec @sarabala1979 have made an initial cut at the changes moving to the config, let me know what you think.

File: config/config.go [hunk 6]
```diff
@@ -106,12 +107,16 @@ type Config struct {
...
-	// Defaults to 30 seconds.
+	// PodGCGracePeriodSeconds specifies the duration in seconds before a terminating pod is forcefully killed.
+	// Value must be non-negative integer. A zero value indicates that the pod will be forcefully terminated immediately.
+	// Defaults to the Kubernetes default of 30 seconds.
 	PodGCGracePeriodSeconds *int64 `json:"podGCGracePeriodSeconds,omitempty"`
 
+	// PodGCDelaySeconds specifies the duration in seconds before the pods in the GC queue get deleted.
+	// Value must be non-negative integer. A zero value indicates that the pods will be deleted immediately.
+	// Defaults to 30 seconds.
+	PodGCDeleteDelay *metav1.Duration `json:"podGCDeleteDelay,omitempty"`
```
Comment (stefansedich) [inline config/config.go:118]: I used a duration here, I was torn between using seconds like the other option above or moving to the more user-friendly duration so went with the more friendly option.

Comment (sarabala1979) [reply config/config.go:118]: Please rename with type `PodGCDeleteDelayDuration`

Comment (stefansedich) [reply config/config.go]: Ok done @sarabala1979

File: config/config.go (hunk 4 above)
Comment (alexec) [reply config/config.go:154]: I think this should be lower - 5s

Comment (stefansedich) [reply config/config.go]: Lowered to 5s by default.

**Review by terrytangyuan — COMMENTED** [review]
I believe you'll have to fix the DCO check.

**Comment by stefansedich** [comment]
> I believe you'll have to fix the DCO check.

Fixed!

**Comment by alexec** [comment]
I've marked this for v3.2, but we may consider back-porting to v3.1.

**Review by alexec — APPROVED** [review]

**Comment by prein** [comment]
Is this meant to have no per-workflow configuration or no configuration at all? I'm not sure, reading the conversation above.
What I would find best for my scenarios would be if I could set different delay for failed and for succeeded workflows. I would then have much longer delay for failed ones to allow for manual inspection.

**Comment by alexec** [comment]
The goal of this PR is to delay all deletion by a minimum amount of time to allow for other services to clean them up. I think you just need normal PodGC.

**Pull request merged**